gen4gl/trunk/fglbld/man/report.man
'\" @(#)$Id: report.man,v 1.1.1.1 2002-06-15 05:07:10 afalout Exp $
'\" @(#)Manual page: General Library -- Report Configuration
.ds fC "@(#)$Id: report.man,v 1.1.1.1 2002-06-15 05:07:10 afalout Exp $
.TH REPORT 3S "JLSS Informix Tools"
.SH NAME
set_output \(em Dynamically reconfigure dimensions of I4GL report
.SH SYNOPSIS
.ft B
function set_plength(i)
.br
define i integer
.sp
function set_tmargin(i)
.br
define i integer
.sp
function set_bmargin(i)
.br
define i integer
.sp
function set_lmargin(i)
.br
define i integer
.sp
function set_rmargin(i)
.br
define i integer
.sp
function set_output()
.ft R
.ft B
function get_plength()
.sp
function get_tmargin()
.sp
function get_bmargin()
.sp
function get_lmargin()
.sp
function get_rmargin()
.ft R
.SH DESCRIPTION
These functions can be used to dynamically reconfigure the layout
of the page of an Informix 4GL report.
The example program shown below illustrates how they should be used.
.br
.ft CW
.ps 10
.vs 12
.nf
.so rejig.4gl
.br
.fi
.ft
.ps
.vs
The functions \fIset_plength\fP, \fIset_tmargin\fP,
\fIset_bmargin\fP, \fIset_lmargin\fP and \fIset_rmargin\fP may be called
anywhere inside an I4GL program.
The function \fIset_output\fP will have no effect except within a report.
It is recommended that \fIset_output\fP be used in the \s-2FIRST
PAGE HEADER\s0 control block, but it may be used in any control block.
.P
After the function \fIset_output\fP is called, it invalidates all the settings;
to change the settings in a second report, you must call the
set-routines again.
There is no requirement that all the routines are called (or even any of them).
.P
The functions \fIget_plength\fP, \fIget_tmargin\fP,
\fIget_bmargin\fP, \fIget_lmargin\fP and \fIget_rmargin\fP may be called
anywhere inside an I4GL program.
If page dimensions have been set using the set-routines, then the corresponding
value is returned; otherwise, the corresponding default value is returned.
After \fIset_output\fP has been called, the values returned will be
the default values again; if your report needs to know what its dimensions will
be, it should use the get-routines before calling \fIset_output\fP.
.SH COMPILATION
If compiling for use with compiled Informix-4GL, no special flags are needed.
If compiling for use with I4GL-RDS, include -DI4GL_RDS on the command line.
The details of building the functions into the \*Cfglgo\*D
and \*Cfgldb\*D are given in the corresponding Informix manuals.
The necessary definitions for \*cfgiusr.c\*d are:
.br
.ft CW
.ps 10
.vs 12
.nf
#ifdef SET_REPORT
extern  int set_output();   /* 0 */
extern  int get_rmargin();  /* 0 */
extern  int get_tmargin();  /* 0 */
extern  int get_bmargin();  /* 0 */
extern  int get_lmargin();  /* 0 */
extern  int get_plength();  /* 0 */
extern  int set_rmargin();  /* 1 */
extern  int set_tmargin();  /* 1 */
extern  int set_bmargin();  /* 1 */
extern  int set_lmargin();  /* 1 */
extern  int set_plength();  /* 1 */
#endif /* SET_REPORT */

#ifdef SET_REPORT
    { "set_output",     set_output,     0   },
    { "get_rmargin",    get_rmargin,    0   },
    { "get_tmargin",    get_tmargin,    0   },
    { "get_bmargin",    get_bmargin,    0   },
    { "get_lmargin",    get_lmargin,    0   },
    { "get_plength",    get_plength,    0   },
    { "set_rmargin",    set_rmargin,    1   },
    { "set_tmargin",    set_tmargin,    1   },
    { "set_bmargin",    set_bmargin,    1   },
    { "set_lmargin",    set_lmargin,    1   },
    { "set_plength",    set_plength,    1   },
#endif /* SET_REPORT */
.br
.fi
.ft
.ps
.vs
.SH WARNINGS
This code was designed and tested on an ATT&T 3B2/400 running 
UNIX System V Release 3.0 using Informix-4GL version 1.10.00C.
It has since been tested on a number of other environments (including Versions
1.10.03, 4.00 and 4.10) on a number of platforms, and has been found to work.
However, the code cannot be guaranteed on any platform or version of I4GL.
It is completely unsupported code.
.SH BUGS
The top margin set by \fBset_tmargin\fP does not take effect
until the second page of output.
.SH AUTHOR
Jonathan Leffler
.br
Marlin Prowell
.br
JLSS
.br
24th July 1992


lib4gl/trunk/libi4gl/src/report.c
head	2.1;
access;
symbols;
locks; strict;
comment	@ * @;


2.1
date	92.07.24.10.16.02;	author johnl;	state Exp;
branches;
next	1.5;

1.5
date	90.10.09.14.22.48;	author john;	state Exp;
branches;
next	1.4;

1.4
date	90.10.05.16.52.58;	author john;	state Exp;
branches;
next	1.3;

1.3
date	90.01.12.14.03.40;	author john;	state Exp;
branches;
next	1.2;

1.2
date	88.04.21.14.28.31;	author john;	state Exp;
branches;
next	1.1;

1.1
date	88.04.20.19.22.26;	author john;	state Exp;
branches;
next	;


desc
@@


2.1
log
@Include Marlin Prowell's mods for I4GL_RDS
@
text
@/*
@@(#) File:           $RCSfile: report.c,v $
@@(#) Version:        $Revision: 1.1 $
@@(#) Last changed:   $Date: 2002-06-14 05:03:47 $
@@(#) Purpose:        I4GL: Reset report output parameters (RDS & C4GL)
@@(#) Author:         J Leffler / M Prowell
*/

/*TABSTOP=4*/

/*
** INFORMIX-4GL Dynamic Report Output Parameters Version 2
** C4GL code (and interface) by J Leffler.
** RDS interface by M Prowell.
**
** Compilation: 
** C4GL: cc -c $RCSfile: report.c,v $
** RDS:  cc -c -DI4GL_RDS $RCSfile: report.c,v $
*/

/*
** Addendum to Marlin's comments.
** -- tested on SUNOS 4.1.1 and I4GL-RDS 4.10.UC1 and it seems to work
**    correctly, but it is still unsupported code.
*/

/*
** From: uunet!nyssa.wa7ipx.ampr.org!mbp (Marlin Prowell)
** Subject: Changing report output params in RDS (code enclosed)
** Date: 29 Apr 92 04:30:46 GMT
** To: informix-list@@rmy.emory.edu
** X-Informix-List-Id: <news.1132>
** 
** A few days ago, Jonathan Leffler posted some code for I4GL that would
** change report output parameters on the fly.  The code he posted works
** only for the C compiler version, and I only have the RDS version of
** Informix.  Since an Informix employee said he was unable to figure out
** a RDS version, I took that as a challenge.
** 
** Below is a modified version of his report.c that works when linked into
** a customized runner.  I leave the details to building the runner as an
** exercise for the reader.
** 
** I used Jonathan's code as a hint on how reports work in Informix.  I
** found the equivalent report pointer, and inferred the report structure
** that the interpreter uses.  Since this code diddles with internal
** interpreter variables, use this code at your own risk.  I use it now to
** reset the page length when output is going to the screen, and it
** *appears* to work correctly.  Your mileage may vary.
** 
** BTW, I am using Informix 4.00 on SCO Xenix.  I haven't tested this
** anywhere else.  Of course, the size of the junk variable in the repdesc
** structure may change on other machine architectures.  You could write
** another C function that gets called in the ON EVERY ROW section, and
** have it print out the values in the repdesc structure.  You can then
** verify that the structure definition is correct on your machine.
** 
** I've wanted to do this, but after I had previously poked around in the
** interpreter for awhile, I had also given up.  Thanks for the hints, Jonathan.
** 
**  -------------------------------------------------------------------------- 
** | Marlin Prowell            | There is a very thin line between ignorance  |
** | (206) 676-1554            | and arrogance and I have totally obliterated |
** | mbp@@nyssa.wa7ipx.ampr.org | that line.             -- Dr. Science        |
**  -------------------------------------------------------------------------- 
*/

#include <stdio.h>

#ifndef lint
static  char    sccs[] = "@@(#)$Id: report.c,v 1.1 2002-06-14 05:03:47 afalout Exp $";
#endif

#ifdef I4GL_RDS

/* Structure inferred from RDS */
typedef struct repdesc
{
	short junk[42];     /* unknown       */
	short ln;           /* Line number?  */
	short pagenumber;   /* Pge number    */
	short plength;      /* Page length */
	short tmargin;      /* Top margin    */
	short bmargin;      /* Bottom margin */
	short lmargin;      /* Left margin   */
	short rmargin;      /* Right margin  */
	short fphlines;     /* Lines in first page header */
	short phlines;      /* Lines in page header */
	short ptlines;      /* Lines in page trailer */
	short junktoo;      /* unknown       */
	short tot;
}  Report;

extern  Report *currep;   /* Parameters of current report */

#else

/* Structure copied from C code generated by I4GL */
typedef struct repdesc
{
	long rrecordnum;    /* Record number */
	short pagenumber;   /* Pge number    */
	short ln;           /* Line number?  */
	short tmargin;      /* Top margin    */
	short rmargin;      /* Right margin  */
	short lmargin;      /* Left margin   */
	short bmargin;      /* Bottom margin */
	short phlines;      /* Lines in page header */
	short fphlines;     /* Lines in first page header */
	short colcount;
	short tot;
	short oktoinc;
	FILE *outfp;        /* Output channel */
	short oftype;
	short plength;      /* Page length */
	short ptlines;      /* Lines in page trailer */
	struct aggdesc *_paggdesc;
	struct value *_paggvals;
	struct sortdesc *p_sortdesc;
	short gotoindx;
	short gotostk[5];
	short needlmarg;
}  Report;

extern  Report *c_rp;   /* Parameters of current report */

#endif	/* I4GL_RDS */

static  Report set;     /* Newly set report parameters  */
static  int newset = 0; /* Mask of newly set parameters */

#define TMARGIN 0x01
#define BMARGIN 0x02
#define LMARGIN 0x04
#define PLENGTH 0x08
#define RMARGIN 0x10

#define DEF_TMARGIN   3
#define DEF_BMARGIN   3
#define DEF_LMARGIN   5
#define DEF_PLENGTH  66
#define DEF_RMARGIN 132

/* Redefine page length */
int set_plength(i)
int i;
{
	if (i == 1)
	{
		popint(&i);
		if (i > 0)
		{
			set.plength = i;
			newset |= PLENGTH;
		}
	}
	return(0);
}

/* Reset top margin */
/* This will not reset the top margin on the first page */
int set_tmargin(i)
int i;
{
	if (i == 1)
	{
		popint(&i);
		if (i >= 0)
		{
			set.tmargin = i;
			newset |= TMARGIN;
		}
	}
	return(0);
}

/* Reset bottom margin */
int set_bmargin(i)
int i;
{
	if (i == 1)
	{
		popint(&i);
		if (i >= 0)
		{
			set.bmargin = i;
			newset |= BMARGIN;
		}
	}
	return(0);
}

/* Reset left margin */
int set_lmargin(i)
int i;
{
	if (i == 1)
	{
		popint(&i);
		if (i >= 0)
		{
			set.lmargin = i;
			newset |= LMARGIN;
		}
	}
	return(0);
}

/* Reset right margin */
int set_rmargin(i)
int i;
{
	if (i == 1)
	{
		popint(&i);
		if (i >= 0)
		{
			set.rmargin = i;
			newset |= RMARGIN;
		}
	}
	return(0);
}

/* Copy reset output list into report configuration */
/* Call in first page header block */
int set_output(i)
int i;
{
	int newlen;
	Report	*report;

#ifdef I4GL_RDS
	report = currep;
#else
	report = c_rp;
#endif	/* I4GL_RDS */

	if (i == 0 && newset && report != (Report *)0)
	{
		if (newset & PLENGTH)
		{
			report->plength = set.plength;
		}
		if (newset & TMARGIN)
		{
			newlen  = report->plength
					- report->ptlines
					- report->phlines
					- report->bmargin
					- report->tmargin
					- 1;
			if (set.tmargin < newlen)
			{
				report->tmargin = set.tmargin;
			}
		}
		if (newset & LMARGIN)
		{
			report->lmargin = set.lmargin;
		}
		if (newset & BMARGIN)
		{
			newlen  = report->plength
					- report->ptlines
					- report->phlines
					- report->tmargin
					- report->bmargin
					- 1;
			if (set.bmargin < newlen)
			{
				report->bmargin = set.bmargin;
			}
		}
		/* Adding report->phlines leaves the report     */
		/* title at the top of the terminal screen.  MP */
		report->tot = report->plength - (report->bmargin + report->ptlines
						+ report->phlines);
		newset = 0;
	}
	return(0);
}

/* Return page length */
int get_plength(i)
int i;
{
	if (newset & PLENGTH)
		i = set.plength;
	else
		i = DEF_PLENGTH;
	retint(i);
	return(1);
}

/* Return top margin */
int get_tmargin(i)
int i;
{
	if (newset & TMARGIN)
		i = set.tmargin;
	else
		i = DEF_TMARGIN;
	retint(i);
	return(1);
}

/* Return bottom margin */
int get_bmargin(i)
int i;
{
	if (newset & BMARGIN)
		i = set.bmargin;
	else
		i = DEF_BMARGIN;
	retint(i);
	return(1);
}

/* Return left margin */
int get_lmargin(i)
int i;
{
	if (newset & LMARGIN)
		i = set.lmargin;
	else
		i = DEF_LMARGIN;
	retint(i);
	return(1);
}

/* Return right margin */
int get_rmargin(i)
int i;
{
	if (newset & RMARGIN)
		i = set.rmargin;
	else
		i = DEF_RMARGIN;
	retint(i);
	return(1);
}
@


1.5
log
@Fix pushint into retint; allow margins to be set to 0
@
text
@d5 61
a65 2
@@(#) Purpose:        I4GL: Reset report output parameters
@@(#) Author:         J Leffler
d74 24
a124 1
#ifndef I4GL_RDS
d126 2
a127 1
#endif /* I4GL_RDS */
d231 7
d239 1
a239 2
#ifndef I4GL_RDS
	if (i == 0 && newset && c_rp != (Report *)0)
d243 1
a243 1
			c_rp->plength = set.plength;
d247 5
a251 5
			newlen  = c_rp->plength
					- c_rp->ptlines
					- c_rp->phlines
					- c_rp->bmargin
					- c_rp->tmargin
d255 1
a255 1
				c_rp->tmargin = set.tmargin;
d260 1
a260 1
			c_rp->lmargin = set.lmargin;
d264 5
a268 5
			newlen  = c_rp->plength
					- c_rp->ptlines
					- c_rp->phlines
					- c_rp->tmargin
					- c_rp->bmargin
d272 1
a272 1
				c_rp->bmargin = set.bmargin;
d275 4
a278 1
		c_rp->tot = c_rp->plength - (c_rp->bmargin + c_rp->ptlines);
a280 1
#endif /* I4GL_RDS */
@


1.4
log
@RDS gets data stored; set_output is still a no-op.
Add get_XXXXXXX routines; include set_rmargin and get_rmargin
@
text
@d85 1
a85 1
		if (i > 0)
d101 1
a101 1
		if (i > 0)
d117 1
a117 1
		if (i > 0)
d133 1
a133 1
		if (i > 0)
d201 1
a201 1
	pushint(i);
d213 1
a213 1
	pushint(i);
d225 1
a225 1
	pushint(i);
d237 1
a237 1
	pushint(i);
d249 1
a249 1
	pushint(i);
@


1.3
log
@Add conditional compilation for guts of routine
@
text
@d44 2
a47 1
#endif /* I4GL_RDS */
d53 7
a64 3
	int j;

#ifndef I4GL_RDS
d67 2
a68 2
		popint(&j);
		if (j > 0)
d70 1
a70 1
			set.plength = j;
a73 1
#endif /* I4GL_RDS */
a81 3
	int j;

#ifndef I4GL_RDS
d84 2
a85 2
		popint(&j);
		if (j > 0)
d87 1
a87 1
			set.tmargin = j;
a90 1
#endif /* I4GL_RDS */
a97 3
	int j;

#ifndef I4GL_RDS
d100 2
a101 2
		popint(&j);
		if (j > 0)
d103 1
a103 1
			set.bmargin = j;
a106 1
#endif /* I4GL_RDS */
d114 11
a124 1
	int j;
d126 4
a129 1
#ifndef I4GL_RDS
d132 2
a133 2
		popint(&j);
		if (j > 0)
d135 2
a136 2
			set.lmargin = j;
			newset |= LMARGIN;
a138 1
#endif /* I4GL_RDS */
d156 1
a156 1
	 if (newset & TMARGIN)
d191 60
@


1.2
log
@detab|ltab; remove debug
@
text
@d42 1
d46 1
d58 2
d69 1
d79 2
d90 1
d99 2
d110 1
d119 2
d130 1
d141 1
d181 1
@


1.1
log
@Initial revision
@
text
@d2 5
a6 5
@@(#)	File:			$RCSfile: report.c,v $
@@(#)	Version:		$Revision: 1.1 $
@@(#)	Last changed:	$Date: 2002-06-14 05:03:47 $
@@(#)	Purpose:		I4GL: Reset report output parameters
@@(#)	Author:			J Leffler
d12 1
a12 1
static	char	sccs[] = "@@(#)$Id: report.c,v 1.1 2002-06-14 05:03:47 afalout Exp $";
d15 1
d18 9
a26 8
	long rrecordnum;
	short pagenumber;
	short ln;
	short tmargin;
	short rmargin;
	short lmargin;
	short bmargin;
	short phlines, fphlines;
d30 1
a30 1
	FILE *outfp;
d32 2
a33 1
	short plength, ptlines;
d42 3
a44 3
extern	Report *c_rp;
static	Report set;
static	int	newset = 0;
d47 1
a47 1
#define BMARGIN	0x02
d52 2
a53 2
int	set_plength(i)
int	i;
d55 1
a55 1
	int	j;
a62 1
			fprintf(stderr, "set plength  to %d\n", j);
d70 2
a71 2
int	set_tmargin(i)
int	i;
d73 1
a73 1
	int	j;
a80 1
			fprintf(stderr, "set tmargin  to %d\n", j);
d87 2
a88 2
int	set_bmargin(i)
int	i;
d90 1
a90 1
	int	j;
a97 1
			fprintf(stderr, "set bmargin  to %d\n", j);
d104 2
a105 2
int	set_lmargin(i)
int	i;
d107 1
a107 1
	int	j;
a114 1
			fprintf(stderr, "set lmargin  to %d\n", j);
d122 2
a123 2
int	set_output(i)
int	i;
d125 1
a125 1
	int	newlen;
a126 1
	fprintf(stderr, "set_output()\n");
a131 1
			fprintf(stderr, "set lmargin  to %d\n", c_rp->plength);
d133 1
a133 1
		if (newset & TMARGIN)
a143 1
				fprintf(stderr, "set lmargin  to %d\n", c_rp->tmargin);
a148 1
			fprintf(stderr, "set lmargin  to %d\n", c_rp->bmargin);
a160 1
				fprintf(stderr, "set bmargin  to %d\n", c_rp->bmargin);
@
